### PR TITLE
[Feature] Add min_area_polygons CUDA op for rotated detection.

### PR DESCRIPTION
**Pull request by zytx121**
Add min_area_polygons CUDA ops for rotated detection.

```
@inproceedings{Guo_2021CVPR_CFA,
  author    = {Zonghao Guo, Chang Liu, Xiaosong Zhang, Jianbin Jiao, Xiangyang Ji and Qixiang Ye},
  title     = {Beyond Bounding-Box: Convex-hull Feature Adaptation for Oriented and Densely Packed Object Detection},
  booktitle = {IEEE/CVF Conference on Computer Vision and Pattern Recognition (CVPR)},
  month     = {June},
  year      = {2021}
}

```

## Modification

Several files in mmcv/ops folder.

## BC-breaking (Optional)

No.

## Use cases (Optional)

from mmcv.ops import min_area_polygons

### CODE REVIEW

**Comment by zhouzaida** [comment]
Please @AllentDan have a look.

File: mmcv/ops/csrc/pytorch/cuda/min_area_polygons.cu [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+// modified from
+// https://github.com/
```
Comment (AllentDan) [inline mmcv/ops/csrc/pytorch/cuda/min_area_polygons.cu:3]: need specified

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+#define MIN_AREA_POLYGONS_CUDA_KERNEL_CUH
+
+#ifdef MMCV_USE_PARROTS
+#include "parrots_cuda_helper.hpp"
+#else
+#include "pytorch_cuda_helper.hpp"
+#endif
+
+#define maxn 20
+const float eps = 1E-8;
```
Comment (AllentDan) [inline mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:12]: may use upper case

File: mmcv/ops/csrc/pytorch/cuda/min_area_polygons.cu (hunk 1 above)
Comment (teamwong111) [inline mmcv/ops/csrc/pytorch/cuda/min_area_polygons.cu:3]: Is there an exact url missing?

**Review by AllentDan — COMMENTED** [review]
Need update ops in https://github.com/open-mmlab/mmcv/blob/master/docs/en/understand_mmcv/ops.md and https://github.com/open-mmlab/mmcv/blob/master/docs/zh_cn/understand_mmcv/ops.md

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+  }
+
+  minpoints[6] = temp[0][0];
+  minpoints[7] = temp[0][1];
+}
+
+__global__ void min_area_polygons_cuda_kernel(const int ex_n_boxes,
+                                              const float *ex_boxes,
+                                              float *minbox) {
+  const int ex_start = blockIdx.x;
```
Comment (teamwong111) [inline mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:371]: Is a "kernel loop" needed here?

File: docs/en/understand_mmcv/ops.md [hunk 4]
```diff
@@ -19,6 +19,7 @@ We implement common CUDA ops used in detection, segmentation, etc.
 - GroupPoints
 - KNN
 - MaskedConv
+- MinAeraPolygon
```
Comment (grimoire) [inline docs/en/understand_mmcv/ops.md:22]: Area?

File: mmcv/ops/min_area_polygons.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    """Find the smallest polygons that surrounds all points in the point sets.
+
+    Args:
+        pointsets (Tensor): point sets with shape  (N, 2 * K), where the
+        integer K > 4.
+
+    Returns:
+        torch.Tensor: Return the smallest polygons with shape (N, 8).
+    """
+    polygons = pointsets.new_zeros((pointsets.size(0) * 8))
```
Comment (grimoire) [inline mmcv/ops/min_area_polygons.py:17]: Why not create a tensor with shape `(pointsets.size(0), 8)` ?

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+      }
+    }
+  }
+}
+
+__device__ inline void Findminbox(float const *const p, float *minpoints) {
+  Point ps1[MAXN];
+  Point convex[MAXN];
+  float pi = 3.1415926;
+  for (int i = 0; i < 9; i++) {
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:277]: In python docstring, pointsets has shape (N, 2 * K) where K>4. But `p` might read wrong value if k<9.

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+  float x, y;
+  __device__ Point() {}
+  __device__ Point(float x, float y) : x(x), y(y) {}
+};
+
+__device__ inline bool point_same(Point &a, Point &b) {
+  return sig(a.x - b.x) == 0 && sig(a.y - b.y) == 0;
+}
+
+__device__ inline void swap1(Point *a, Point *b) {
```
Comment (AllentDan) [inline mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:28]: Do we have some common implementation of functions like swap for `__device__` in the common folder? @zhouzaida

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh (hunk 3 above)
Comment (zytx121) [reply mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh]: Done.

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+    }
+  }
+  if (max_index == 0) {
+    max_index = 1;
+    p_max = in_poly[max_index];
+  }
+
+  k_index = 0, Stack[0] = 0, top1 = 0;
+  while (k_index != max_index) {
+    p_k = p_max;
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:212]: could use `cross` here?

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+  }
+
+  for (int i = 0; i <= top1; i++) {
+    right_point[i] = in_poly[Stack[i]];
+  }
+
+  k_index = 0, Stack[0] = 0, top2 = 0;
+
+  while (k_index != max_index) {
+    p_k = p_max;
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:236]: could use `cross` here?

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh (hunk 6 above)
Comment (zytx121) [reply mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:280]: You are right.

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+  float R[2][2];
+
+  R[0][0] = cos(angle);
+  R[0][1] = cos(angle - pi / 2);
+  R[1][0] = cos(angle + pi / 2);
+  R[1][1] = cos(angle);
+
+  float temp[1][2];
+  float points[1][2];
+  points[0][0] = xmax;
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:306]: Can you merge these 4 nested loops? Or use 1d array for temp and points.

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+    if (i <= top1) {
+      in_poly[i] = right_point[i];
+    } else {
+      in_poly[i] = left_point[top2 - (i - top1)];
+    }
+  }
+  n_poly = top1 + top2;
+  for (int i = 0; i < n_poly; i++) {
+    for (int j = 0; j < n_input; j++) {
+      if (point_same(in_poly[i], input_poly[j])) {
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:263]: It seems that `points_to_convex_ind` has not been used. Can we ignore these code if `points_to_convex_ind` is nullptr?

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh (hunk 8 above)
Comment (zytx121) [reply mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:207]: Thank you for your kind proposal!

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh (hunk 11 above)
Comment (zytx121) [reply mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:263]: YES!!  `points_to_convex_ind` is really not used. I will delete it directly.

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh (hunk 10 above)
Comment (zytx121) [reply mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh]: Done.

File: mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh (hunk 7 above)
Comment (zytx121) [reply mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:21]: This function is for `Point`, which doesn't have a common implementation in mmcv as far as I know.

Comment (AllentDan) [reply mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh:21]: OK

File: mmcv/ops/csrc/pytorch/cuda/min_area_polygons.cu [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+#include "min_area_polygons_cuda.cuh"
+#include "pytorch_cuda_helper.hpp"
+
+void MinAreaPolygonsCUDAKernelLauncher(const Tensor pointsets,
+                                       Tensor polygons) {
+  int num_pointsets = pointsets.size(0);
+  const int output_size = polygons.numel();
+  at::cuda::CUDAGuard device_guard(pointsets.device());
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  min_area_polygons_cuda_kernel<<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0,
```
Comment (zhouzaida) [inline mmcv/ops/csrc/pytorch/cuda/min_area_polygons.cu:13]: Should we use `AT_DISPATCH_FLOATING_TYPES_AND_HALF`? @grimoire

Comment (grimoire) [reply mmcv/ops/csrc/pytorch/cuda/min_area_polygons.cu:13]: Yes, I think so.

Comment (zytx121) [reply mmcv/ops/csrc/pytorch/cuda/min_area_polygons.cu]: Done.

**Comment by zhouzaida** [comment]
Please resolve the conflicts.

**Review by grimoire — APPROVED** [review]
LGTM

**Review by zhouzaida — APPROVED** [review]

**Review by teamwong111 — APPROVED** [review]

**Review by AllentDan — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**